### PR TITLE
Feature: Enable/Disable kerberos authentication for cloudera manager hbase instance

### PR DESCRIPTION
**Pull request by VenkatasivareddyTR**
*Issue #, if available:*
Kerberos authentication is not supported for HBase connector.

*Description of changes:*
Added changes to handle Kerberos authentication for Cloudera manager HBase instance.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java [hunk 1]
```diff
@@ -116,6 +126,39 @@ private HBaseConnection createConnection(String host, String masterPort, String
...
+            if (StringUtils.isNotBlank(s3uri)) {
+                try {
+                    Path tempDir = copyConfigFilesFromS3ToTempFolder(configOptions);
+                    logger.debug("tempDir: " + tempDir);
+                    keytabLocation = tempDir + File.separator + "hbase.keytab";
+                    System.setProperty("java.security.krb5.conf", tempDir + File.separator + "krb5.conf");
+                    logger.debug("keytabLocation: " + keytabLocation);
+                }
+                catch (Exception e) {
+                    logger.error("Exception while copying config files from S3 to temp folder: " + e.getMessage());
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java:146]: generally speaking you want to either log or throw an exception;

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java [hunk 2]
```diff
@@ -116,6 +126,39 @@ private HBaseConnection createConnection(String host, String masterPort, String
...
+                    throw new RuntimeException(e);
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java:147]: This is where you could send up the error; and using the constructor that allows you send a message and the reason the exception was thrown; see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java [hunk 3]
```diff
@@ -116,6 +126,39 @@ private HBaseConnection createConnection(String host, String masterPort, String
...
+                }
+            }
+
+            UserGroupInformation.setConfiguration(config);
+            try {
+                String principalName = configOptions.get(PRINCIPAL_NAME);
+                UserGroupInformation.loginUserFromKeytab(principalName, keytabLocation);
+            }
+            catch (IOException ex) {
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java:156]: same as above;

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java [hunk 4]
```diff
@@ -88,6 +98,41 @@ private Configuration getHbaseConfiguration(String connectionStr)
...
+            String s3uri = configOptions.get(KERBEROS_CONFIG_FILES_S3_REFERENCE);
+            if (StringUtils.isNotBlank(s3uri)) {
+                try {
+                    Path tempDir = copyConfigFilesFromS3ToTempFolder(configOptions);
+                    logger.debug("tempDir: " + tempDir);
+                    keytabLocation = tempDir + File.separator + "hbase.keytab";
+                    System.setProperty("java.security.krb5.conf", tempDir + File.separator + "krb5.conf");
+                    logger.debug("krb5.conf location: " + tempDir + File.separator + "krb5.conf");
+                }
+                catch (Exception e) {
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:117]: same as above

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java [hunk 5]
```diff
@@ -88,6 +98,41 @@ private Configuration getHbaseConfiguration(String connectionStr)
...
+                }
+            }
+            logger.debug("keytabLocation: " + keytabLocation);
+
+            UserGroupInformation.setConfiguration(configuration);
+            try {
+                String principalName = configOptions.get(PRINCIPAL_NAME);
+                UserGroupInformation.loginUserFromKeytab(principalName, keytabLocation);
+            }
+            catch (IOException ex) {
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:129]: ++++

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java (hunk 1 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java]: Hi @AbdulR3hman, updated in latest commit.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java (hunk 2 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java]: Hi @AbdulR3hman, updated in latest commit.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 4 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:117]: Hi @AbdulR3hman, Code has been updated in latest commit.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 5 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:128]: Hi @AbdulR3hman, updated in latest commit.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java [hunk 6]
```diff
@@ -113,6 +123,40 @@ private Connection createConnection(String host, String masterPort, String zooke
...
+                if (StringUtils.isNotBlank(s3uri)) {
+                    try {
+                        Path tempDir = copyConfigFilesFromS3ToTempFolder(configOptions);
+                        logger.debug("tempDir: " + tempDir);
+                        keytabLocation = tempDir + File.separator + "hbase.keytab";
+                        System.setProperty("java.security.krb5.conf", tempDir + File.separator + "krb5.conf");
+                        logger.debug("krb5.conf location: " + tempDir + File.separator + "krb5.conf");
+                    }
+                    catch (Exception e) {
+                        throw new RuntimeException("Exception while copying config files from S3 to temp folder: " + e);
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java:144]: Still not using the right connector; see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java [hunk 7]
```diff
@@ -113,6 +123,40 @@ private Connection createConnection(String host, String masterPort, String zooke
...
+                }
+                logger.debug("keytabLocation: " + keytabLocation);
+
+                UserGroupInformation.setConfiguration(config);
+                try {
+                    String principalName = configOptions.get(PRINCIPAL_NAME);
+                    UserGroupInformation.loginUserFromKeytab(principalName, keytabLocation);
+                }
+                catch (IOException ex) {
+                    throw new RuntimeException("Exception in UserGroupInformation.loginUserFromKeytab: " + ex);
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java:155]: same here; see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseKerberosUtils.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+     * Gets the environment variable.
+     *
+     * @param key - the config key
+     * @return {@link String}
+     */
+    private static String getRequiredConfig(String key, java.util.Map<String, String> configOptions)
+    {
+        String value = configOptions.getOrDefault(key, "");
+        if (value.isEmpty()) {
+            throw new IllegalArgumentException("Lambda Environment Variable " + key + " has not been populated! ");
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseKerberosUtils.java:120]: see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java [hunk 9]
```diff
@@ -116,6 +126,37 @@ private HBaseConnection createConnection(String host, String masterPort, String
...
+            if (StringUtils.isNotBlank(s3uri)) {
+                try {
+                    Path tempDir = copyConfigFilesFromS3ToTempFolder(configOptions);
+                    logger.debug("tempDir: " + tempDir);
+                    keytabLocation = tempDir + File.separator + "hbase.keytab";
+                    System.setProperty("java.security.krb5.conf", tempDir + File.separator + "krb5.conf");
+                    logger.debug("keytabLocation: " + keytabLocation);
+                }
+                catch (Exception e) {
+                    throw new RuntimeException("Exception while copying config files from S3 to temp folder: " + e);
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java:146]: see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java [hunk 10]
```diff
@@ -116,6 +126,37 @@ private HBaseConnection createConnection(String host, String masterPort, String
...
+                }
+            }
+
+            UserGroupInformation.setConfiguration(config);
+            try {
+                String principalName = configOptions.get(PRINCIPAL_NAME);
+                UserGroupInformation.loginUserFromKeytab(principalName, keytabLocation);
+            }
+            catch (IOException ex) {
+                throw new RuntimeException("Exception in UserGroupInformation.loginUserFromKeytab: " + ex);
```
Comment (AbdulR3hman) [inline athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java:156]: see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 5 above)
Comment (AbdulR3hman) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:128]: see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 4 above)
Comment (AbdulR3hman) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:117]: see [`public RuntimeException(String message,Throwable cause)`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-)

**Review by AbdulR3hman — COMMENTED** [review]
please review it again.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java (hunk 10 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java (hunk 9 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/connection/HbaseConnectionFactory.java]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseKerberosUtils.java (hunk 8 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseKerberosUtils.java:120]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java (hunk 7 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java (hunk 6 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 5 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:128]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 4 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:117]: Hi @AbdulR3hman, Updated code as per your suggestion.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java [hunk 11]
```diff
@@ -113,6 +123,40 @@ private Connection createConnection(String host, String masterPort, String zooke
...
+                if (StringUtils.isNotBlank(s3uri)) {
+                    try {
+                        Path tempDir = copyConfigFilesFromS3ToTempFolder(configOptions);
+                        logger.debug("tempDir: " + tempDir);
+                        keytabLocation = tempDir + File.separator + "hbase.keytab";
+                        System.setProperty("java.security.krb5.conf", tempDir + File.separator + "krb5.conf");
+                        logger.debug("krb5.conf location: " + tempDir + File.separator + "krb5.conf");
+                    }
+                    catch (Exception e) {
+                        throw new RuntimeException("Exception while copying config files from S3 to temp folder: " + e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseConnectionFactory.java]: Thank you @AbdulR3hman, Updated the error message and removed concatenation.

File: athena-hbase/athena-hbase.yaml [hunk 12]
```diff
@@ -91,6 +111,15 @@ Resources:
...
                 - s3:ListAllMyBuckets
+                - s3:ListBucket
+                - s3:GetObject
+                - s3:GetBucketLocation
+                - s3:GetObjectVersion
+                - s3:PutObject
+                - s3:PutObjectAcl
+                - s3:GetLifecycleConfiguration
+                - s3:PutLifecycleConfiguration
+                - s3:DeleteObject
```
[the remark this answers is not in the document]
Comment (VenkatasivareddyTR) [reply athena-hbase/athena-hbase.yaml]: Thank you @AbdulR3hman, Updated code with necessary permissions.

File: athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java (hunk 4 above)
Comment (VenkatasivareddyTR) [reply athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/integ/HbaseTableUtils.java:117]: Hi @AbdulR3hman Thank you, Updated code as per your suggestion.

**Comment by VenkatasivareddyTR** [comment]
Hi @AbdulR3hman, I have addressed your review comments, if you have some time can you please have a look at once.

**Review by AbdulR3hman — APPROVED** [review]

**Pull request merged**